### ISSUE TITLE
Delete and Update subquery support

### ISSUE THREAD

**Issue by tdas** [delta-io/delta#826]
`DELETE FROM myTable1 WHERE EXISTS (Select 1 from myTable2 where myTable2.key = myTable1.key)`

This query will currently throw an error, "Subqueries are not supported in the DELETE" 
We want to support subqueries in Delete and Update.

**Comment by FidelCastillo** [comment]
I have the same error with something like

`UPDATE Table1
SET Column1 = 1
WHERE Column2 IN (SELECT Column2 FROM Table2)`

**Comment by tdas** [comment]
Yeah. Its the same issue. We did not support it earlier because some earlier version of Spark did not have the right internal support for filter with subqueries (Delete/Update implementation needs filters with subqueries). I think this limiation has been resolved in later versions of Spark, so we should try removing this line and see whether it works - https://github.com/delta-io/delta/blob/master/core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableDelete.scala#L36 

Anybody willing to contribute can test this out.

**Comment by FidelCastillo** [comment]
no work with spark 3.1.2

**Comment by zevektor** [comment]
Hi, 

I was attempting to perform something similar but I got the error that subqueries, as well as window and aggregate functions cannot be used in a DELETE statement. 

It would be quite useful to have this kind of support, for use cases like deleting old versions of the same record:

Suppose I have an append-only delta table, it would be extremely useful to be able to do something like this in order to get rid of older record versions and keep the table tidy:

```scala
deltaTable.delete(row_number().over(Window.partitionBy(primary_key).orderBy(desc(timestamp_column))) === 1)
```

or a subquery:

```sql
with records_to_delete as (
SELECT key, timestamp
FROM deltatable
WHERE row_number() over (partition by key) order by timestamp desc = 1
)
DELETE FROM TABLE deltatable WHERE (key, timestamp) in (select key, timestamp from records_to_delete)
```

I was pointed out to use the merge statement as a workaround, but I think using delete is a more straightforward approach and would be good to have these kinds of statement supported by delete commands.

**Comment by mantovani** [comment]
When this going to be fixed ?

**Comment by akshaymahajans** [comment]
bump ^ - when will this be fixed?

**Comment by wassimtrifi** [comment]
+1

**Comment by GP202301** [comment]
+1

**Comment by mantovani** [comment]
Is this ever be fixed ?

**Comment by HuyTCM** [comment]
Up, we really want to have the subqueries in Delete and Update conditions.

**Comment by TawabG** [comment]
+1

**Comment by aleksandraangelova** [comment]
Ran into this issue today and would also really appreciate if nested queries are supported in DELETE clause.

**Comment by rasskazovp** [comment]
+1

**Comment by LazyRen** [comment]
It's been 5 years and yet we do not have a solution for this... 👎🏻

**Comment by felipepessoto** [comment]
#6541 - [Spark] Support DELETE with EXISTS deterministic source subquery
#6543 - Open-source and add DELETE deterministic source subquery EXISTS tests

**Comment by foss-contributor** [comment]
@tdas, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!

**Comment by felipepessoto** [comment]
Still valid, and some folks are actively working on it
